parsing: Attach a location to the RHS of Ptyp_alias

Julow opened the pull request:
Hi,

This patch is part of OCamlformat's [extended parser](https://github.com/ocaml-ppx/ocamlformat/blob/main/vendor/parser-extended/parsetree.mli), which is OCaml's parser with modifications to add missing locations and represent the different concrete syntaxes.

Merging some of these modifications back would help us keep the two parsers in sync but I think would also improve the parser for other uses (eg. error messages, PPX, etc..).

In this case, the change improves an error message.

Julow committed:
parsing: Attach a location to the RHS of Ptyp_alias

Julow committed:
typing: Use more precise location of alias type mismatch

Before:

    1 | type t = (int as 'a) * (float as 'a)
                                ^^^^^^^^^^^
    Error: This alias is bound to type float but is used as an instance of type
             int

After:

    1 | type t = (int as 'a) * (float as 'a)
                                          ^
    Error: This alias is bound to type float but is used as an instance of type
             int

Julow commented on testsuite/tests/typing-external/pr11392.ml:
```diff
@@ -16,9 +16,9 @@ type 'self nat = Z | S of 'self
+Line 1, characters 37-41:
 1 | external cast : int -> 'self nat as 'self = "%identity"
-                    ^^^^^^^^^^^^^^^^^^^^^^^^^
+                                         ^^^^
```
I find it weird that the location does not span the quote but this is the same as in let-bindings. 

I propose to change the location to include the quote. The quote is not included in the AST but is logically part of the ident as suggested by this error message:

```
1 | fun (x : 'a t as 'a) -> ();;
                      ^
Error: This alias is bound to type "'a t" but is used as an instance of type "'a"
       The type variable "'a" occurs inside "'a t"
```

Octachron replied:
I tend to agree that it would be better to include the quote in the location.

Octachron commented on testsuite/tests/typing-misc/typetexp_errors.ml:
```diff
@@ -36,9 +36,9 @@ Error: The type variable name "'_a" is not allowed in programs
 1 | let f (x: int as 'a) (y: float as 'a) = (x,y)
-                             ^^^^^^^^^^^
+                                       ^
 Error: This type "float" should be an instance of type "int"
```
There is a mismatch between the error message (which focuses on the types) and the highlighted location in this case.

Julow replied:
Fixed.

Octachron commented on typing/typetexp.ml:
```diff
@@ -526,22 +526,22 @@ and transl_type_aux env ~row_context ~aliased ~policy styp =
             with_local_level_if_principal begin fun () ->
               let t = newvar () in
-              TyVarEnv.remember_used alias t styp.ptyp_loc;
+              TyVarEnv.remember_used alias.txt t alias.loc;
```
I propose to keep `styp.ptyp_loc` in this case: this would move the location from `Type_mismatch` in
```
let f (x: int as 'a) (y:float as 'a) = ...
```
to 
```ocaml
(y:float as 'a)
```
> Error: This type float should be an instance of type int

rather than just the alias, while only somehow degrading the quality of error messages in the pathological case
```ocaml
type t = 'a as 'b
```
(the location move from the `'b` alias to `'a as 'b` )

Julow replied:
Thanks, fixed.

Octachron commented on parsing/ast_helper.ml:
```diff
@@ -103,9 +103,9 @@ module Typ = struct
-            check_variable var_names t.ptyp_loc string;
-            Ptyp_alias(loop core_type, string)
+        | Ptyp_alias(core_type, alias) ->
+            check_variable var_names t.ptyp_loc alias.txt;
```
The variable is really the alias here, so `t.ptyp_loc` ⇒ `alias.loc`. If you want to test the change in the error message, this case is triggered by:
```ocaml
let none: type a. (_ as 'a) option = None
```

Julow replied:
Fixed in a new commit.

Octachron commented:
Overall, I like the change: it improves a handful of error messages and it makes the parsetree more precise on a meaningful location. I have few fine tuning remarks left however.

Julow committed:
typing: Improve location for 'variable in scope'

Use the location of the alias instead of the `_ as 'a` type expression:

    2 | let none: type a. (_ as 'a) option = None
                                 ^
    Error: In this scoped type, variable 'a is reserved for the local type a.

Co-authored-by: Florian Angeletti <florian.angeletti@inria.fr>

Julow committed:
Update Changes

Octachron commented:
The current state looks good to me and I don't see foresee much problems for ppxlib, cc @panglesd ?

Octachron commented:
Merged after checking with @panglesd  and @pitag-ha that the change should be fine for ppxlib.
